samcli/cli/cli_config_file.py

```diff
@@ -0,0 +1,152 @@
+DEFAULT_IDENTIFER = "default"
+
+
+class TomlProvider:
+    """
+    A parser for toml configuration files
+    :param cmd: sam command name as defined by click
+    :param section: section defined in the configuration file nested within `cmd`
+    """
+
+    def __init__(self, cmd=None, section=None):
+        self.cmd = cmd
+        self.section = section
+
+    def __call__(self, file_path, identifier, cmd_name):
+        """
+        :param file_path: The path to the configuration file
+        :param identifier: The name of the sectional identifier within configuration file.
+        :param cmd_name: sam command name as defined by click
+        :returns dictionary containing the configuration parameters under specified identifier
+        """
+        config = toml.load(file_path)
+        resolved_config = {}
+        if self.section:
+            try:
+                resolved_config = self._get_identifier(config, identifier)[self.cmd][self.section]
+            except KeyError:
+                LOG.debug(
+                    "Error reading configuration file at %s with identifier %s, command %s, section %s",
+                    file_path,
+                    identifier,
+                    self.cmd,
+                    self.section,
+                )
+        return resolved_config
+
+    def _get_identifier(self, config, identifier):
+        """
+
+        :param config: loaded TOML configuration file into dictionary representation
+        :param identifier: top level section defined within TOML configuration file
+        :return:
+        """
+        return config.get(identifier)
+
+
+def configuration_callback(cmd_name, option_name, identifier_name, saved_callback, provider, ctx, param, value):
+    """
+    Callback for reading the config file.
+
+    Also takes care of calling user specified custom callback afterwards.
+
+    :param cmd_name: `sam` command name derived from click.
+    :param option_name: The name of the option. This is used for error messages.
+    :param identifier_name: `top` level section within configuration file
+    :param saved_callback: User-specified callback to be called later.
+    :param provider:  A callable that parses the configuration file and returns a dictionary
+        of the configuration parameters. Will be called as
+        `provider(file_path, identifier, cmd_name)`.
+    :param ctx: Click context
+    :param param: Click parameter
+    :param value: Specified value for identifier
+    :returns specified callback or the specified value for identifier.
+    """
+    ctx.default_map = ctx.default_map or {}
+    cmd_name = cmd_name or ctx.info_name
+
+    config_file = os.getenv("SAM_CONFIG", os.path.abspath(os.path.join(".", DEFAULT_CONFIG_FILE_NAME)))
+
+    param.default = DEFAULT_IDENTIFER
+    identifier_name = value or identifier_name
+
+    if os.path.isfile(config_file):
+        try:
+            LOG.debug("Config file location: %s", os.path.abspath(config_file))
```

Comment:
so this `os.path.abspath` again here is dangerous. It is a good practice to do relative->absolute path conversion once because it's tricky getting the right cwd